geoh5py/shared/merging/base.py

```diff
@@ -0,0 +1,200 @@
+                id_ += 1
+
+            # define how to define the entity_type
+            if nb_keys != 3 and len(
+                list({unitype[:nb_keys] for unitype in entity_unique_entity_types})
+            ) != len(entity_unique_entity_types):
+                nb_keys = 2
+                if len(list(set(entity_unique_entity_types))) != len(
+                    entity_unique_entity_types
+                ):
+                    nb_keys = 3
+
+            previous += cast(int, input_entity.n_vertices)
+
+        # create an intermediate array
+        all_data = np.array(all_data, dtype=object)
+
+        # prepare the attributes
+        unique_entity_types = list(
+            dict.fromkeys(
+                [entity_type[:nb_keys] for entity_type in unique_entity_types]  # type: ignore
+            )
+        )
+        unique_associations: dict = {
+            entity_type: [] for entity_type in unique_entity_types
+        }
+        data_array = np.empty((len(unique_entity_types), previous))
+        data_array[:] = np.nan
+
+        # populate the intermediate array
+        for data in all_data:
+            data_array[
+                unique_entity_types.index(data[1][:nb_keys]), data[-2] : data[-1]
+            ] = data[0]
+            unique_associations[data[1][:nb_keys]].append(data[2])
+
+        # create the output data
+        add_data_dict = {}
+        for id_, data in enumerate(data_array):
+            # define the association
+            association = unique_associations[unique_entity_types[id_]]
+
+            if not len(set(association)) == 1:
+                raise ValueError(
+                    f"Cannot merge data with different associations: {set(association)}"
+                )
+
+            # define the name
+            if nb_keys == 3:
+                name = f"merged_{unique_entity_types[id_][0].name}_{id_}"
+            elif nb_keys == 2:
+                name = f"merged_{unique_entity_types[id_][0].name}_{unique_entity_types[id_][1]}"
+            else:
+                name = f"merged_{unique_entity_types[id_][0].name}"
+
+            add_data_dict[name] = {
+                "association": association[0],
+                "values": data,
+                "entity_type": unique_entity_types[id_][0],
+            }
+
+        return add_data_dict
+
+    @classmethod
+    def merge_objects(
+        cls, input_entities: list[ObjectBase], add_data: bool = True, **kwargs
+    ):
+        """
+        Merge a list of :obj:geoh5py.objects.ObjectBase of the same type.
+        :param input_entities: the list of objects to merge.
+        :param add_data: if True, the data will be merged.
+        :param kwargs: the kwargs to create the new object.
+        :return: an object of type input_entities[0] containing the merged vertices.
+        """
+        cls.validate_objects(input_entities)
+
+        kwargs = cls.validate_kwargs(input_entities[0].workspace, **kwargs)
```

Comment:
Not sure that's very safe. Technically objects could come from different sources. Feels like the outgoing workspace should be an explicit property of the class. You won't be able to make everything classmethod, but it will be less ambiguous about where things come and go.

Reply:
workspace is defined as an input of merge_object